plat/*: Add support for more modern PCI devices

jschlumpp opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

This introduces support for allocating IRQ numbers and setting up MSI-X tables to use these IRQs. It also introduces support for mapping PCI BAR registers, which are used to do memory-mapped I/O with devices.
<!--
Please provide a detailed description of the changes made in this new PR.
-->

jschlumpp committed:
drivers/ukbus/pci: Basic PCI capability decoding

This is a reduced variant of [1] and allows an iteration over the
capabilities of PCI devices.

[1]: https://github.com/astrynzha/virtiofs_unikraft/commit/7c3e99760fc6751f81e70e785af7ccbee17cabdf.

Co-authored-by: Andrii Strynzha <a.strynzha@gmail.com>
Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
drivers/ukbus/pci: Basic MSI-X configuration interface

MSI-X allow setting up interrupts in a more flexible way without having
multiple devices on the same interrupt line.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
drivers/ukbus/pci: Add helper for writing to PCI conf space

This adds a write helper for PCI register with an API similar to the
read macros.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
drivers/ukbus/pci: Implement helpers to map/unmap BARs

In some cases the BIOS does not already map these and then we
need to map them ourselves.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

mogasergiu committed:
drivers/ukbus/pci: Move `PCI` generic r/w methods to common header

Since the `pci_generic_config_*` function signatures apply to the
`PCI` spec regardless of the configuration mechanism, move their
signatures to a platform common header inclusion location.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
drivers/ukbus/pci: Add `pci_generic_config_*` for `x86` `port I/O`

Implement PCI Configuration Address Space Mechanism reading/writing
`pci_generic_config_*` methods, for the legacy `x86` `port I/O`
mechanism.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

kha-dinh commented on drivers/ukbus/pci/pci_bus.c:
```diff
@@ -158,6 +167,273 @@ void _pci_register_driver(struct pci_driver *drv)
+	rc = ukplat_page_map(pt, bar_virt, bar_phys, bar_pages, attr, 0);
+	if (unlikely(rc)) {
+		pt->fa->ffree(pt->fa, bar_phys, bar_pages);
+		return ENOMEM;
```
Shouldn't `rc` be returned here instead? I don't think `ENOMEM` is appropriate here.

jschlumpp committed:
drivers/ukintctlr/xpic: Configure spurious vector

This ensures that the LAPIC works correctly.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

jschlumpp committed:
plat/common/x86: Initialize APIC when enabled

When the APIC is enabled via KConfig we always want to initialize
it, too. Note that a configured APIC can co-exist with a
configured PIC.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

kha-dinh commented on drivers/ukbus/pci/arch/x86_64/msix.c:
```diff
@@ -0,0 +1,35 @@
+{
+	memset(entry, 0, sizeof(*entry));
+	entry->addr_low =
+		0xfeeU << 20 |
```
I think we can make `0xfeeU` into a macro for better readability.

kha-dinh commented on drivers/ukbus/pci/include/uk/bus/pci.h:
```diff
@@ -298,8 +470,39 @@ static struct pci_bus_handler ph __unused;
+int pci_generic_config_write(__u8 bus, __u8 devfn,
+			     int where, int size, __u32 val);
+/**
+ * @brief Create a memory mapping for a memory BAR
```
I don't see the `@breif` tag used often. Is it a remnant of an old convention?

RaduNichita replied:
I think it should be removed as well

@mschlumpp can you adapt the comments to the [coding-conventions](https://unikraft.org/docs/contributing/coding-conventions#function-documentation), please?

jschlumpp replied:
Ah yeah sure, it's just pretty old code that predates the coding conventions.

kha-dinh commented on drivers/ukbus/pci/pci_bus.c:
```diff
@@ -158,6 +167,272 @@ void _pci_register_driver(struct pci_driver *drv)
+
+	UK_ASSERT(dev);
+	UK_ASSERT(bar_out);
+	UK_ASSERT(idx < 6);
```
In https://github.com/unikraft/unikraft/pull/1124/commits/171bf6bc4da8192c511746c061e08d6765f3655e I made a macro for maximum BAR; maybe it can be done in this PR?

kha-dinh reviewed (commented):
Overall it looks good. I only have some minor comments.

RaduNichita commented on drivers/ukbus/pci/arch/x86_64/pci_bus.c:
```diff
@@ -65,13 +65,71 @@
+	switch (size) {
+	case 8:
+		outb(PCI_CONFIG_DATA + (where & 0x03), val);
+
```
I think we can remove the newline before break

RaduNichita commented on drivers/ukbus/pci/arch/x86_64/pci_bus.c:
```diff
@@ -65,13 +65,71 @@
+
+		break;
+	case 16:
+		outw(PCI_CONFIG_DATA + (where & 0x03), val);
```
can we put a define for 0x03?

jschlumpp replied:
I'm not sure what a good constant name would be. @mogasergiu

RaduNichita commented on drivers/ukbus/pci/arch/x86_64/pci_bus.c:
```diff
@@ -94,9 +152,15 @@ static inline int pci_driver_add_device(struct pci_driver *drv,
 	PCI_CONF_READ(__u16, &dev->base, config_addr, IOBAR);
 	PCI_CONF_READ(__u8, &dev->irq, config_addr, IRQ);
 
+	if (arch_pci_find_cap(dev, PCI_CAP_MSIX, &cap_offset) == 0)
```
Can you add some comment regarding this code? I find it difficult to follow the change without a comment explaining the reason behind.

RaduNichita commented on drivers/ukbus/pci/msix.c:
```diff
@@ -0,0 +1,165 @@
+{
+	uint32_t msix_table_size, msix_bar, msix_table_offset;
+	char *base;
+	volatile struct pci_msix_table_entry *msix_table;
```
Out of curiosity: why these variable have to be marked as volatile?

jschlumpp replied:
I'll replace it with a more explicit barrier.

RaduNichita reviewed (commented):
Hi @mschlumpp ,

I did a first pass on this PR and left a couple of comments.

kha-dinh committed:
drivers/ukbus/pci: Define macro for maximum number of BARs

Changed places where BAR index is checked to use the macro instead
of the hard-coded value 6.

Signed-off-by: Kha Dinh <khadinh@g.skku.edu>
Signed-off-by: Marco Schlumpp <marco@unikraft.io>

kha-dinh commented on drivers/ukbus/pci/arch/x86_64/pci_bus.c:
```diff
@@ -239,3 +301,93 @@ int arch_pci_probe(struct uk_alloc *pha)
 	return 0;
 }
+
+
```
There is an extra new line here.

kha-dinh commented on drivers/ukbus/pci/include/uk/bus/pci.h:
```diff
@@ -150,6 +151,16 @@ enum pci_device_state {
 };
 
+/**
+ * @brief Description of a mapped BAR.
```
There is still a `@brief` tag here.

kha-dinh commented on drivers/ukbus/pci/include/uk/bus/pci.h:
```diff
@@ -150,6 +151,16 @@ enum pci_device_state {
+struct pci_bar_memory {
+	/**< Start address of memory region */
+	void *start;
+	/**< Size of the memory region */
```
I think `*<` is used for putting comments *after* the variable in Doxygen (although I saw a few places where it is used this way in the codebase).

kha-dinh commented on drivers/ukbus/pci/include/uk/bus/pci.h:
```diff
@@ -298,8 +492,49 @@ static struct pci_bus_handler ph __unused;
+
+int pci_generic_config_write(__u8 bus, __u8 devfn,
+			     int where, int size, __u32 val);
+/**
```
Add a newline here?

kha-dinh commented on drivers/ukbus/pci/pci_bus.c:
```diff
@@ -158,6 +167,272 @@ void _pci_register_driver(struct pci_driver *drv)
+		return rc;
+	}
+#else
+	(void)attr;
```
Maybe `__maybe_unused` annotation should be used instead of this access here.

kha-dinh reviewed (commented):
I did a second pass and left some comments.
